libtransmission/blocklist.cc

```diff
@@ -49,10 +50,67 @@ void BlocklistFile::ensureLoaded() const
         return;
     }
 
-    auto range = IPv4Range{};
-    while (in.read(reinterpret_cast<char*>(&range), sizeof(range)))
+    auto file_size = std::filesystem::file_size(filename_);
+    auto zeroes_count = 0;
+    auto max_zeroes = 0;
+
+    if (file_size >= 40)
+    {
+        std::array<char, 40> first_struct = {};
+
+        in.read(reinterpret_cast<char*>(&first_struct), 40);
+        in.clear();
+        in.seekg(0, std::ios::beg);
+
+        for (auto const struct_byte : first_struct)
+        {
+            if (struct_byte == 0)
+            {
+                zeroes_count++;
+            }
+            else
+            {
+                if (zeroes_count > max_zeroes)
+                {
+                    max_zeroes = zeroes_count;
+                }
+
+                zeroes_count = 0;
+            }
+        }
+
+        if (zeroes_count > max_zeroes)
+        {
+            max_zeroes = zeroes_count;
+        }
+    }
+
+    // Check for old blocklist file format
+    // Old struct size was 8 bytes (2 IPv4), new struct size is 40 bytes (2 IPv4, 2 IPv6)
+    //
+    // If we encounter less than 4 continuous bytes containing 0 we are using old file format
+    // (as the new format guarantees at least 2 empty IPv4 OR 2 empty IPv6)
+    // If we confirm using old style convert to new style and rewrite blocklist file
+    if ((file_size >= 40 && max_zeroes < 4) || (file_size % 8 == 0 && file_size % 40 != 0))
     {
-        rules_.emplace_back(range);
+        auto range = AddressRange{};
+        while (in.read(reinterpret_cast<char*>(&range), 8))
+        {
+            rules_.emplace_back(range);
+        }
+
+        tr_logAddInfo(_("Rewriting old blocklist file format to new format"));
```

Comment:
I'm concerned about the possibility of misidentifying the file format.

Maybe instead of having this test logic, could we rename the newly-formatted binary file?

The filename is currently `${basename}.bin`, maybe the new format could be saved in `${basename}6.bin`?

This logic is currently in `tr_session::loadBlocklists()`. It probably ought to be migrated into `blocklist.cc`; whichever makes this PR easier to do

Reply:
I don't think there is a possibility of misidentifying the file, outside of malformed files, which won't be parsed properly anyways.

Creating a new file would entail a lot more checks for different files and such, which I'm to do with but don't see the point in as of now.

The loadBlocklists call I'm happy to change but for scope creep reasons I'd like to do it within another PR